lib/blockchain/chain.js

```diff
@@ -950,6 +956,46 @@ class Chain extends AsyncEmitter {
     return this.emitAsync('reorganize', tip, competitor);
   }
 
+  /**
+   * Checks if a reorganization breaks the set nlrLimit,
+   * if it does, it invalidates necessary blocks and returns true
+   * else return false
+   * @param {ChainEntry} tip - Current tip of this chain.
+   * @param {ChainEntry} fork - The tip of the fork.
+   * @param {ChainEntry} competitor - The competing chain's tip.
+   * @returns {Promise<boolean>}
+   */
+  async noLongReorg (tip, fork, competitor) {
+    if (tip.height - fork.height >= this.network.nlrLimit) {
+      if (this.network.nlrLimit !== 0) {
+        this.logger.warning(
+          'NLR Activated: current=%h(%d) fork=%h(%d) reorg_size=%d',
+          tip.hash,
+          tip.height,
+          competitor.hash,
+          competitor.height,
+          tip.height - fork.height
+        );
+
+        // mark invalid_child from tip of fork to second block of fork
+        while (competitor.height > fork.height + 2) {
+          const previous = await this.getPrevious(competitor)
+          this.invalidate(previous.hash)
+        }
+
+        // mark invalid first block of fork
+        const previous = await this.getPrevious(competitor)
+        this.invalidate(previous.hash)
```

Comment:
Why is this invalidation not a part of the above loop?

Reply:
I was following the pattern implemented in flocore. I don't know what it's done this way

https://github.com/floblockchain/flo/commit/2fe045194caf8137878128ebf01461b9a99e7fb4#diff-24efdb00bfbe56b140fb006b562cc70bR2338

Reply:
Ah, in flocore they're being marked different types of invalid.

The first block of the fork is marked as a "failed block" and all subsequent blocks are marked "failed child" it doesn't look like fcoin distinguishes between the two so the separation seemed odd at first impressions

Reply:
makes sense. I'll change the while loop to just run down the whole thing. 
and am I wrong in thinking this?

```
while (pindexWalk->nHeight != pindexFork->nHeight+2) {
            pindexWalk = pindexWalk->pprev;
```

the first two lines of the invalidation while loop start with taking the `pprev` of the `pindexWalk` and invalidating from there. what about `pindexWalk` itself. Doesn't that have to get invalidated (the tip)? Doesn't it seem like were skipping that by starting with `pindexWalk -> prev`